pylabrobot/resources/plate_adapter.py

```diff
@@ -0,0 +1,175 @@
+    self.adapter_hole_size_y = adapter_hole_size_y
+    self.adapter_hole_size_z = self.get_size_z() - self.dz
+    self.adapter_hole_dx = adapter_hole_dx
+    self.adapter_hole_dy = adapter_hole_dy
+    self.site_pedestal_z = site_pedestal_z
+
+  @property
+  def child_resource_location(self) -> Optional[Coordinate]:
+    return self._child_resource_location
+
+  def assign_child_resource(
+    self,
+    resource: Plate,
+    location: Optional[Coordinate] = None,
+    reassign: bool = True
+  ):
+    """Assign a Plate to a PlateAdapter. If `location` is not provided,
+    the resource will autoadjust the placement location based on the
+    PlateAdapter-Plate relationship."""
+
+    if self._child_resource is not None and not reassign:
+      raise ValueError(f"{self.name} already has a child resource assigned")
+    if not isinstance(resource, Plate):
+      raise ValueError("Only plates can be assigned to Alpaqua 96 magnum flx.")
+
+    # TODO: have discussion oon whether to transfer flat bottom error checking
+    # TODO: check whether allPlate children information could
+    # be made accessible from the Plate class
+
+    # Calculate Plate information (which is not directly accessible from the Plate class)
+    x_locations = sorted(OrderedDict.fromkeys([well_n.location.x
+      for well_n in resource.children]))
+    y_locations = sorted(OrderedDict.fromkeys([well_n.location.y
+      for well_n in resource.children]))
+
+    def calculate_well_spacing(float_list: List[float]):
+      """ Calculate the difference between every x and x+1 element in the list of floats. """
+      if len(float_list) < 2:
+        return None
+      differences = [round(float_list[i+1] - float_list[i],2) for i in range(len(float_list) - 1)]
+      if differences[0] is None:
+        raise ValueError("well spacing has to be uniform, and cannot be None")
+      elif len(list(OrderedDict.fromkeys(differences))) == 1:
+        return differences[0]
+      else:
+        raise ValueError("well spacing has to be uniform")
+
+    plate_dx, plate_dy = float(x_locations[0]), float(y_locations[0])
+    plate_item_dx = abs(calculate_well_spacing(x_locations))
+    plate_item_dy = abs(calculate_well_spacing(y_locations))
+    well_size_x = resource.children[0].get_size_x()
+    well_size_y = resource.children[0].get_size_y()
+    # true_dz = resource.get_size_z() - resource.children[0].get_size_z()
+
+    # Well positioning check
+    assert (plate_item_dx, plate_item_dy) in [(9.0, 9.0), (4.5, 4.5), (2.25, 2.25)], \
+      "PlateAdapter only accepts plates with a well spacing of " + \
+      "9x9, 4.5x4.5 or 2.25x2.25mm^2 (ANSI SLAS 4-2004 (R2012): Well Positions)"
+
+    # Calculate adjustment to place center of H1_plate on top of center of H1_adapter
+    plate_x_adjustment = self.dx - plate_dx + self.adapter_hole_size_x/2 - well_size_x/2
+    plate_y_adjustment = self.dy - plate_dy + self.adapter_hole_size_x/2 - well_size_y/2
```

Comment:
isn't it always true that `self.adapter_hole_size_x == well_size_x`?

Reply:
No, it is not.

1) `well_size_x` does not include the side thickness of the well - the same issue that is breaking PLR's Plate definition in the z dimension, missing information regarding the well_material_thickness (just that in x-y it doesn't matter for aspiration/dispensation)

2) the `self.adapter_hole_size_x` can be significantly different from `well_size_x`:
2.i. it can be much smaller, e.g. in the magnetic rack -> then only a part of the well actually enters the PlateAdapter.hole with the remaining well sticking out of it
2.ii. it can be bigger (e.g. by the thickness of the well material, as mentioned in point 1, exemplified by the Hamilton TCC plate adapter)

That was one of the reasons I added two very different plate adapters in this PR, to showcase the variety of PlateAdapters that are covered by this new class :)

Reply:
Thanks for the explanation!